File: pws.cpp
// 101 Thousand Guesses Per Second
#include <iostream>
#include <string>
#include <vector>
#include <chrono>

using namespace std;

// Function to generate combinations
void generateCombinations(const string& charSet, int maxLength, const string& password, int& guesses) {
    vector<char> currentCombination(maxLength);
    string currentString;

    auto start = chrono::steady_clock::now();

    while (currentString != password) {
        currentString = "";
        for (int i = 0; i < maxLength; ++i) {
            currentString += charSet[currentCombination[i]];
        }

        // Increment the combination
        int pos = maxLength - 1;
        while (pos >= 0 && ++currentCombination[pos] == charSet.size()) {
            currentCombination[pos--] = 0;
        }

        cout << currentString << "\n";

        ++guesses;

        // Check if current combination matches the password string
        if (currentString == password) {
            auto end = chrono::steady_clock::now();
            chrono::duration<double> elapsed_seconds = end - start;
            cout << "Found " << password << " with " << guesses << " guesses" << " in " << elapsed_seconds.count() << " seconds" << endl;
            return;
        }
    }
}

int main() {
    string charSet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890!@#$%^&*()-_=+[]{}|,',.<>?~";
    string password;
    int maxLength;

    cout << "Password: ";
    cin >> password;

    cout << "Length: ";
    cin >> maxLength;

    int guesses = 0;
    generateCombinations(charSet, maxLength, password, guesses);

    return 0;
}

File: python/pws.py
import time

def generate_combinations(char_set, max_length, password):
    char_set_len = len(char_set)
    current_combination = [0] * max_length
    guesses = 0

    start_time = time.time()

    while True:
        guesses += 1
        # Generate the current string from current_combination
        current_string = ''.join(char_set[i] for i in current_combination)


        if current_string == password:
            elapsed_time = time.time() - start_time
            print(f"Found {password} with {guesses} guesses in {elapsed_time:.6f} seconds")
            return

        # Increment the combination
        pos = max_length - 1
        while pos >= 0:
            current_combination[pos] += 1
            if current_combination[pos] == char_set_len:
                current_combination[pos] = 0
                pos -= 1
            else:
                break

if __name__ == "__main__":
    char_set = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890!@#$%^&*()-_=+[]{}|,'.<>?~"
    password = input("Password: ")
    max_length = int(input("Length: "))

    generate_combinations(char_set, max_length, password)
